src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+    let fnGet = (obj:any, prop:string) => {
+      let path = obj.path || [];
+      path.push(prop);
+      let neue:any = () => {};
+      neue.path = path;
+      return new Proxy(neue, {
+        get: fnGet,
+        apply: (target:any, targetThis:any, args:any[]) => {
+          let func = new DSLFunction(this, path, args);
+          this.functions.push(func);
+          return func;
+        }});
+    }
+    return new Proxy({}, {get:fnGet});
+  }
+
+  find = (...args:any[]) => {
+    let lastArg = args[args.length - 1];
+    let proxied:any = {};
+    let tag;
+    if(typeof lastArg === "object") {
+      proxied = lastArg;
+      tag = args.slice(0, args.length - 1);
+    } else {
+      tag = args.slice(0, args.length);
+    }
+    // console.log("tag", tag);
+    let rec = new DSLRecord(this, tag, proxied);
+    this.records.push(rec);
+    return rec.proxy();
+  }
+
+  record = (...args:any[]) => {
+    let out = this.find.apply(null, args);
+    out.output = true;
+    return out;
+  }
+
+  registerVariable(variable:DSLVariable) {
+    this.variableLookup[variable.id] = [variable];
+  }
+
+  toValue(a:any) {
+    if(a === undefined || a === null) throw new Error("Eve values can't be undefined or null");
+    if(a instanceof DSLVariable) {
+      return a.value;
+    } if(a instanceof DSLRecord) {
+      return a.record.value;
+    } if(a instanceof DSLFunction) {
+      return a.returnValue.value;
+    }
+    return a;
+  }
+
+  equivalence(a:any, b:any) {
+    let aValue = a.value || (a.record && a.record.value) || a;
+    let bValue = b.value || (b.record && b.record.value) || b;
+    let aIsRegister = isRegister(aValue);
+    let bIsRegister = isRegister(bValue);
+    if(aIsRegister && bIsRegister) {
+      let aVars = this.variableLookup[a.id];
+      let bVars = this.variableLookup[b.id];
+      for(let variable of aVars) {
+        variable.value = bValue;
+        bVars.push(variable);
+      }
+      this.variableLookup[a.id] = [];
+    } else if(aIsRegister) {
+      let aVars = this.variableLookup[a.id];
+      for(let variable of aVars) {
+        variable.value = bValue;
+      }
+      this.variableLookup[a.id] = [];
+    } else if(bIsRegister) {
+      let bVars = this.variableLookup[b.id];
+      for(let variable of bVars) {
+        variable.value = aValue;
+      }
+      this.variableLookup[b.id] = [];
+    } else if(aValue !== bValue) {
+      throw new Error(`Trying to equivalence two static values that aren't the same: ${aValue} and ${bValue}`);
+    }
+  }
+
+  prepare() {
+    let functions = this.functions.slice() as (DSLFunction | undefined)[];
+    // look through the functions for equivalences before we do anything else.
+    let ix = 0;
+    for(let func of functions) {
+      if(!func || func.path[1] !== "==") continue;
+      this.equivalence(func.args[0], func.args[1]);
+      functions[ix] = undefined;
+      ix++;
+    }
+    let registerIx = 0;
+    for(let id in this.variableLookup) {
```

Comment:
I see, so it really will be a register, we just haven't assigned it yet. Let's maybe add a constant for UNASSIGNED_REG or something?